logger.go

```diff
@@ -0,0 +1,175 @@
+func (l Logger) InfoWithDepth(extraDepth int, args ...interface{}) {
+	l.printWithDepth(infoLog, extraDepth, l.extendWithPfx(args)...)
+}
+
+// Infoln is equivalent to the global Infoln function, with the addition of prefix and data content from this Logger.
+func (l Logger) Infoln(args ...interface{}) {
+	l.println(infoLog, l.extendWithPfx(args)...)
+}
+
+// InfolnWithDepth is equivalent to the global InfolnWithDepth function, with the addition of prefix and data content from this Logger.
+func (l Logger) InfolnWithDepth(extraDepth int, args ...interface{}) {
+	l.printlnWithDepth(infoLog, extraDepth, l.extendWithPfx(args)...)
+}
+
+// Infof is equivalent to the global Infof function, with the addition of prefix and data content from this Logger.
+func (l Logger) Infof(format string, args ...interface{}) {
+	l.printf(infoLog, l.pfx(format), l.extend(args)...)
+}
+
+// InfofWithDepth is equivalent to the global InfofWithSepth function, with the addition of prefix and data content from this Logger.
+func (l Logger) InfofWithDepth(extraDepth int, format string, args ...interface{}) {
+	l.printfWithDepth(infoLog, extraDepth, l.pfx(format), l.extend(args)...)
+}
+
+// Warning is equivalent to the global Warning function, with the addition of prefix and data content from this Logger.
+func (l Logger) Warning(args ...interface{}) {
+	l.print(warningLog, l.extendWithPfx(args)...)
+}
+
+// WarningWithDepth is equivalent to the global WarningWithDepth function, with the addition of prefix and data content from this Logger.
+func (l Logger) WarningWithDepth(extraDepth int, args ...interface{}) {
+	l.printWithDepth(warningLog, extraDepth, l.extendWithPfx(args)...)
+}
+
+// Warningln is equivalent to the global Warningln function, with the addition of prefix and data content from this Logger.
+func (l Logger) Warningln(args ...interface{}) {
+	l.println(warningLog, l.extendWithPfx(args)...)
+}
+
+// WarninglnWithDepth is equivalent to the global WarninglnWithDepth function, with the addition of prefix and data content from this Logger.
+func (l Logger) WarninglnWithDepth(extraDepth int, args ...interface{}) {
+	l.printlnWithDepth(warningLog, extraDepth, l.extendWithPfx(args)...)
+}
+
+// Warningf is equivalent to the global Warningf function, with the addition of prefix and data content from this Logger.
+func (l Logger) Warningf(format string, args ...interface{}) {
+	l.printf(warningLog, l.pfx(format), l.extend(args)...)
+}
+
+// WarningfWithDepth is equivalent to the global WarningfWithDepth function, with the addition of prefix and data content from this Logger.
+func (l Logger) WarningfWithDepth(extraDepth int, format string, args ...interface{}) {
+	l.printfWithDepth(warningLog, extraDepth, l.pfx(format), l.extend(args)...)
+}
+
+// Error is equivalent to the global Error function, with the addition of prefix and data content from this Logger.
+func (l Logger) Error(args ...interface{}) {
```

Comment:
In the error methods, I would also add an option for actually passing the err param explicitly. I am not sure about proper naming, though. Perhaps we can do a WithError?

Reply:
I'd like to avoid changing the pattern of the logging functions in this change.

Reply:
Well, the whole point is to improve the logging situation...
I would love to be able to do something like:
logger := glog.WithPrefix("abracadabra")
logger.Info("here comes the magic")
err := doMagic()
if err != nil {
  logger.WithError(err).Error("no magic for you!")
}

problem is, the error will stick around. Need to think a bit more about this :)

Reply:
This is probably a case of scope creep. But definitely something we could take as a separate PR/CR.

Reply:
I disagree this is scope creep as we are building the next generation of logging for our system. There is no point in limiting ourselves to only the features we have now. Especially something as widespread as logging. As it is, we will now have 3 ways of doing logging (log4go, glog, the new logger). If we make changes at a later stage, we will have 4.
I do think this could be useful:
1. takes away the guesswork as to where to put the error in the param list
2. spares the endlessly repeating Errorf("...error: %v", ..., err) and reduces the bugs that have to do with positioning the vars in the wrong order
3. more strongly typed: at least the error is guaranteed to be type checked
4. uniform error formatting (as it will be concatenated inside the logger implementation)

I think that a builder pattern, similar to WithPrefix and WithData could be beneficial: logger.WithError(err) + glog.WithError(err); the only downside is we have to remember not to re-assign the result of logger.WithError(err) back to the logger var; but since it's the same exact thing with prefix and data, I feel it's not that bad.

Reply:
This is not an additional way to do logging. This is still glog, it's just providing convenience methods so it's clearer that adding data is an option, and make some of our code easier to read. 

    glog.WithData(map[string]interface{}{"id": operationId}).Info("We did something")

Is the pattern I foresee.

This actually reduces the number of ways of doing logging, as we also have publishcontext as a wrapper around glog with context.Context just for Pages publishes, which as it turns out is way too specialized.

The Error builder is an addition on top of this pattern, and not one we need to make right this minute. We don't even know that this pattern will work or be comfortable yet. Let's get something actually in use, and then we can see where it will go.

Reply:
Let's say I'm convinced ;)
But your example brought me to another thought: is there a way to simplify the bulky map[string]interface{} each time? Perhaps define a type glog.Data for that?

Reply:
Off the top of my head, I think that's something we'd need to deal with in the glog-contrib/raven package.

Reply:
Not necessarily. We currently define the interface between our code and the glog logger. What happens inside glog is not our concern. For now, we can even convert it back to map[string]interface{} when passing down the stack. As long as the contract between our code and glog is concise, I'm happy.

Reply:
No, I mean that I think the need for a map is a quirk of Sentry, not of glog. The glog-contrib/raven package is used to handle the backend to forward to Sentry, which performs checks on types, expecting a map, http request or one of the raven structs for selecting a dsn or fingerprint.

Reply:
What happens when we don't pass a map then?
Either way, it makes sense to pass key-value pairs as data in some cases, even if it's not required (e.g. the default data map in publishcontext). When that happens, we can use glog.Map instead of map[string]interface{}. When that's not the case, we can do whatever we do now.
Also, your response made me think of another thing: WithFingerprint(args ...interface{}) ? 
So a call like this (alpha/gocode/src/yext/pages/publishing/templating/data.go:957):
glog.Error(
  errgo.Newf(
    "unknown choice ID %v for multi-option custom field \"%v\" (ID: %v)",
    choice,
    cf.GetName(),
    cf.GetId(),
  ),
  glog.Data(publishcontext.GetDataWithFeatureInfo(ctx)),
  glog.Data(raven.Fingerprint("unknown choice", fmt.Sprint(cf.GetId()))),
)

Could look like this:
glog.WithData(
  publishcontext.GetDataWithFeatureInfo(ctx)
).WithFingerprint(
    "unknown choice", cf.GetId()
  ).Errorf(
      "unknown choice ID %v for multi-option custom field \"%v\" (ID: %v)",
      choice,
      cf.GetName(),
      cf.GetId(),
    )

Looks much more structured and elegant (I'm cheating a bit, of course, as the original call creates an unnecessary error instead of doing glog.Errorf directly; but still...)

And if the error would be defined previously, I would love to see something like this:
err := errgo.Newf(
    "unknown choice ID %v for multi-option custom field \"%v\" (ID: %v)",
    choice,
    cf.GetName(),
    cf.GetId(),
)
glog.WithData(
  publishcontext.GetDataWithFeatureInfo(ctx)
).WithFingerprint(
    "unknown choice", cf.GetId()
  ).WithError(
      err
    ).Error()

Reply:
(I worked so hard to make the formatting right and they just trimmed it when I submitted :(  It looks much better with indentation)

Reply:
glog.WithData(
publishcontext.GetDataWithFeatureInfo(ctx)
).WithFingerprint(...)

Is mixing functionality from different projects.
Fingerprint is Sentry specific, and added by glog-contrib, it's totally separate from this package.

Reply:
Well, technically, data is also a senty thing. We just use glog to pass it along (without really caring about what's in it) :) 
On the one hand, typing  .WithData(raven.Fingerprint("unknown choice", fmt.Sprint(cf.GetId()))) every time is annoying. On the other, creating yet another logger wrapper that encapsulates both this one and raven is also not great (although it can be easily achieved by embedding this one and adding the missing parts). So yeah, I don't know. I agree that importing raven from glog is out of the question. But I'm not entirely sure that the fingerprint concept cannot be generalized (e.g. every log message, regardless of the backend, can have more important and less important components).
After all, the whole point of this is convenience. No new functionality is introduced.

Reply:
The concept of a backend is totally removed from Sentry. And Fingerprinting is very much only needed for something that "rolls up" data. There would be no need for this for backends like Logstash.